### PR TITLE
External app form download request fixes & additions

### LINKED ISSUE #2462: Fix ODK Collect form download request intent from external app
#### Software and hardware versions 
Collect v1.16.1, Android v6.0, 
Devices Used: Infinix, Oppo

#### Problem description

Service intents do not work on the above phones. These are the only tested phones on which this behaviour has been replicated so far. 

#### Steps to reproduce the problem

1. User opens ODK Collect so that it is kept in memory
1. User opens ODK Collet Intents Tester app - Download the app from [here](https://github.com/opendatakit/collect/files/2167038/app-debug.apk.zip)
1. User opens DOWNLOAD FORM
1. User adds FORM_ID and clicks on DOWNLOAD FORM button

#### Expected behavior
Form with added id is downloaded and visible in ODK Collect

#### Other information 
I tried keeping ODK Colect in memory, checking the stack trace, testing the behaviour when the phone had enough charge and was not in battery saver mode.

### PR DESCRIPTION
**Pull request by ekigamba**
Closes #2462 

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lintDebug spotbugsDebug` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

Changes made:
- [x] Change service to broadcast receiver
- [x] Authenticate automatically using saved url, username & password when this is required
- [x] Enable the use of optional url, username & password
- [x] Enable form download via foreground for cases where the external app cannot communicate in the background with ODK Collect. This is by enabling direct downloads from the *Get Blank Form* activity

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/receivers/FormDownloadRequestReceiver.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import timber.log.Timber;
+
+public class FormDownloadRequestReceiver extends BroadcastReceiver {
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        if (intent.hasExtra(ApplicationConstants.BundleKeys.FORM_ID) && intent.hasExtra(ApplicationConstants.BundleKeys.TRANSACTION_ID)) {
+            processRequest(intent);
+        }
+        // This is a dev error and not a user error ?? Not sure if I should return a response. It should fail silently
```
Comment (jd-alexander) [inline collect_app/src/main/java/org/odk/collect/android/receivers/FormDownloadRequestReceiver.java]: Yeah, a simple Timber log would do.

Comment (ekigamba) [reply collect_app/src/main/java/org/odk/collect/android/receivers/FormDownloadRequestReceiver.java]: Cool, didn't think of that

**Comment by ekigamba** [comment]
> Check if form with the similar ID exists (This is to avoid a situation where the user cannot open saved forms because the parent form was replaced)

@ln @yanokwa Could we assume that its a form update since this doesn't make the previous saved instance useless, the previous form is still retained. In this case this should work OK for the foreground form download which extends the functionality in FormDownloadList activity. 

It seems that even the background download handles form updates the same way foreground download does on FormDownloadList activity from [here](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/utilities/FormDownloader.java#L325)

Form updates/"replacement" is not available if the forms have similar md5 hashes, the code deletes the new downloaded form and retains the previous form.

I don't know but would two forms ever have the same md5 hash?

So, I tested it out and it seems that the forms are in both cases(foreground and backgroudn download) not replaced, the only issue would be that the user might lose access to filling in the previous version of the form.

It works to the advantage of the third party if the form update is still downloaded so that the third parties can control form updates. The third party forms cannot be updated automatically by ODK Collect since ODK Collect doesn't hold the aggregate server settings.

**Comment by lognaturel** [comment]
@ekigamba You are correct that blank forms are never actually replaced by an update, they're just hidden in the UI.

> the user might lose access to filling in the previous version of the form.

When would it happen? The reason all blank forms are retained is to avoid this case.

MD5 hash collisions are really unlikely and I don't think you need to consider that case.

Can you please say more about the case you're trying to guard against? Maybe describing the user/client app action and what should be triggered on the Collect side?

A couple of high-level comments:
- The current plan is to target a 9/14 code freeze for a 9/23 release in case you want to try to get this in
- Like in #2461, it would be good if the username and password extras were optional. Either way, it should be consistent.
- Since the Service wasn't yet documented, I think it's ok to remove/replace entirely in this next release.

**Comment by ekigamba** [comment]
Oh! 🤔 In the initial discussion of the feature, the possibility of partially filled forms being inaccessible came out as one of the dangers and it seems clear that it wouldn't happen.

> the only issue would be that the user might lose access to filling in the previous version of the form.

The user might not be able to fill in a previous blank form version since the recently downloaded version is the one which is displayed in the **Fill Blank Form** page.

> Like in #2461, it would be good if the username and password extras were optional. Either way, it should be consistent.

I will definitely borrow from #2461 

> Since the Service wasn't yet documented, I think it's ok to remove/replace entirely in this next release.

👍 I will remove this. Added it to the current PR tasks

> The current plan is to target a 9/14 code freeze for a 9/23 release in case you want to try to get this in

Thanks, I actually want this in by then.

**Comment by lognaturel** [comment]
> In the initial discussion of the feature, the possibility of partially filled forms being inaccessible came out as one of the dangers

I now remember @yanokwa bringing up some questions about replacing blank forms during the call we had to talk about this and I see notes [here](https://docs.google.com/document/d/1hszoTRzWG5W04JXgcBzE7BcdEZGB75lA8tftTNZ5lzA/edit#). There is some risk in letting an external application replace a blank form but this can already be done directly through shared storage manipulation or changing things through the content providers so it doesn't seem like it adds any new risks.

As for partially filled forms, I think there is only one case where one would become inaccessible and if it's a form with the same `form_id` and `version` **but different contents** were downloaded. Manual downloads have the same issue so I don't think it should be addressed here.

> The user might not be able to fill in a previous blank form version since the recently downloaded version is the one which is displayed in the Fill Blank Form page.

What's the use case for this? It's really important that a user be able to edit a partially-filled form, that I agree with. But I can't think of a case where a user filling in a new form would need to fill a previous version. The way I see it, if two versions are relevant at the same time, they should really be considered different forms with different `form_id`s. Is this a requirement for you?

**Comment by ekigamba** [comment]
> Is this a requirement for you?

Nope.

**Comment by ekigamba** [comment]
@lognaturel I made the required changes:

- Removed the service and broadcast receiver
- Made the username & password optional for anonymous logins

File: collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java [hunk 2]
```diff
@@ -61,24 +66,51 @@ public DownloadFormListUtils() {
...
+        SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(
                         Collect.getInstance().getBaseContext());
-        String downloadListUrl =
+
+        // Remove trailing '/'
+        while (url != null && url.endsWith("/")) {
+            url = url.substring(0, url.length() - 1);
+        }
+
+        //Todo: Make the username/password optional for cases of anonymous logins
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java]: I think this is addressed?

File: collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java [hunk 3]
```diff
@@ -61,24 +66,51 @@ public DownloadFormListUtils() {
...
+        String downloadListUrl = url != null ? url :
                 settings.getString(PreferenceKeys.KEY_SERVER_URL,
                         Collect.getInstance().getString(R.string.default_server_url));
         // NOTE: /formlist must not be translated! It is the well-known path on the server.
         String formListUrl = Collect.getInstance().getApplicationContext().getString(
                 R.string.default_odk_formlist);
-        String downloadPath = settings.getString(PreferenceKeys.KEY_FORMLIST_URL, formListUrl);
+
+        // When a url is supplied, we will use the default formList url
+        String downloadPath = (url != null && username != null && password != null) ? formListUrl : settings.getString(PreferenceKeys.KEY_FORMLIST_URL, formListUrl);
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java]: This part should just be `url != null` now.

**Comment by lognaturel** [comment]
Couple of small comments inline. Once you address them (if I'm wrong, just give a brief explanation), please comment with `@opendatakit-bot "needs testing"` so it can go in the QA queue! I've verified the basic behavior with https://github.com/grzesiek2010/collectTester/pull/19/files. Would also be good to get a regression check around downloading blank forms.

I /think/ it will be easier to merge #2461 first but let me know if you have a preferred order.

File: collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java (hunk 2 above)
Comment (ekigamba) [reply collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java]: Yea

File: collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java (hunk 3 above)
Comment (ekigamba) [reply collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java]: On it

**Comment by ekigamba** [comment]
> Would also be good to get a regression check around downloading blank forms.

Right

> I /think/ it will be easier to merge #2461 first but let me know if you have a preferred order.

I also prefer this

**Comment by ekigamba** [comment]
@opendatakit-bot "needs testing"

**Comment by lognaturel** [comment]
Thanks, @ekigamba! @mmarciniak90, the last commit makes a change more related to #2461 -- it changes the way the instance upload is accessed. You'll see it's just a manifest change so I'm not sure there's much to verify other than you'll see https://github.com/grzesiek2010/collectTester/pull/22 was also updated to match.

**Comment by mmarciniak90** [comment]
Hi @ekigamba 
I have a problem with downloading form via FORM_ID in Tester app.
`FORM [added id]: Download requested` status is visible and it does not change.
I created the apk from your branch with the newest changes.
Tester app was downloaded from https://github.com/opendatakit/collect/issues/2462#issue-349021144

**Comment by ekigamba** [comment]
@mmarciniak90 That's the old APK, let me build one and post it here

**Comment by ekigamba** [comment]
[Here](https://github.com/opendatakit/collect/files/2383256/app-debug.apk.zip) is the latest tester app apk. This apk only helps test the form download

**Comment by mmarciniak90** [comment]
Thanks, @ekigamba!

@lognaturel, I can confirm that it works in that way on all tested devices:

tester

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Verified cases:
* Downloading form with media
* Downloading the same form few times
* Incorect FORM_ID
* No internet connection


* Incorrect url


* Aggregate does not allow anonymous user
* Cancel downloading


@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

**Review by lognaturel — APPROVED** [review]

**Pull request merged**